tests/lang/s02_core/s07_type_inference/t01_return_type_inference.nim

```diff
@@ -0,0 +1,36 @@
+discard """
+  description: '''
+    The specified return type of a non-generic routines may be generic, in
+    which case the concrete type is inferred from the body.
+  '''
+"""
+
+## Using 'auto' as the return type enables unconstrained return type
+## inference. The real return type is taken from the source expression
+## of the first analysed assignment to the `result` variable.
+
+proc p(): auto =
+  result = 1
+
+doAssert typeof(p()) is int
+
+## The return type specified on the procedure definition may also be a
+## type-class of a built-in generic type, in which case it acts as a
+## constraint.
+
+proc p2(): tuple =
```

Comment:
```suggestion
proc p2(): tuple =
  ## equivalent to `proc p2[R: tuple](): R =`
```

nice to have, just read slightly more spec like.

Reply:
They're not equivalent, as `proc p2[R: tuple](): R` requires the call to specify the type at the callsite (no return type inference would be used then) whereas for `proc p2(): tuple` this is neither required nor possible.

The original changes included that `p2` wouldn't be a generic routine in the first place (it doesn't have to be one), but that seemed like something better changed separately.

Reply:
Ah, because it's in the signature so inference based on the body wouldn't happen.